Add a new config object for compactConfig

jihoonson opened the pull request:
Fixes #5242.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/5264)
<!-- Reviewable:end -->

leventov commented on server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java:
```diff
@@ -21,158 +21,127 @@
+  @JsonCreator
+  public CoordinatorCompactionConfig(
+      @JsonProperty("compactionConfigs") List<DataSourceCompactionConfig> compactionConfigs,
+      @JsonProperty("compactionTaskSlotRatio") Double compactionTaskSlotRatio,
```
Annotate `@Nullable`

jihoonson replied:
Added.

leventov commented on server/src/main/java/io/druid/server/coordinator/CoordinatorCompactionConfig.java:
```diff
@@ -21,158 +21,127 @@
+  public CoordinatorCompactionConfig(
+      @JsonProperty("compactionConfigs") List<DataSourceCompactionConfig> compactionConfigs,
+      @JsonProperty("compactionTaskSlotRatio") Double compactionTaskSlotRatio,
+      @JsonProperty("maxCompactionTaskSlots") Integer maxCompactionTaskSlots
```
And this

jihoonson replied:
Added.

leventov commented on server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java:
```diff
@@ -331,8 +294,6 @@ public static Builder builder()
-    private static final double DEFAULT_COMPACTION_TASK_RATIO = 0.1;
-    private static final int DEFAILT_MAX_COMPACTION_TASK_SLOTS = Integer.MAX_VALUE;
 
     private Long millisToWaitBeforeDeleting;
```
Please annotate properties and parameters in this class `@Nullable` as appropriate.

jihoonson replied:
Added.

leventov commented on server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -0,0 +1,178 @@
+
+public class DataSourceCompactionConfig
+{
+  // should be synchronized with Tasks.DEFAULT_MERGE_TASK_PRIORITY
```
Please make this a javadoc comment with a link

jihoonson replied:
`Tasks` is in a different module and can't be linked.

leventov commented on server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -0,0 +1,178 @@
+  private final String dataSource;
+  private final int taskPriority;
+  private final long targetCompactionSizeBytes;
+  // The number of compaction segments is limited because the byte size of a serialized task spec is limited by
```
And here

jihoonson replied:
Same. Can't be linked.

leventov commented on server/src/main/java/io/druid/server/coordinator/DataSourceCompactionConfig.java:
```diff
@@ -0,0 +1,178 @@
+  @JsonCreator
+  public DataSourceCompactionConfig(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("taskPriority") Integer taskPriority,
```
`@Nullable`, and 3 params after this

jihoonson replied:
Added.

nishantmonu51 commented on server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java:
```diff
@@ -0,0 +1,209 @@
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
```
minor nit: is it possible to add some message about the failure cause here ? Maybe something on lines - Failed to set config...

jihoonson replied:
Added.

nishantmonu51 commented on server/src/main/java/io/druid/server/http/CoordinatorCompactionConfigsResource.java:
```diff
@@ -0,0 +1,209 @@
+        newCompactionConfig,
+        new AuditInfo(author, comment, req.getRemoteAddr())
+    )) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
```
minor nit: same, please add more details for error if possible.

jihoonson replied:
Added.

gianm commented:
Marking as 0.13.0 since #5242 is marked as such; and this patch is responsive to that issue.

jihoonson commented:
@leventov @nishantmonu51 @gianm thanks for the review. Addressed your comments.

@gianm there are already some docs. Here are the links.

- https://github.com/druid-io/druid/blob/master/docs/content/design/coordinator.md#compacting-segments
- https://github.com/druid-io/druid/blob/master/docs/content/design/coordinator.md#http-endpoints

Do you think something more should be added?

gianm commented:
@jihoonson, doesn't this patch add a new endpoint: `/druid/coordinator/v1/config/compaction` and remove the compaction config from the dynamic config (where it used to be)? I meant reflecting those changes in the docs.

jihoonson commented:
@gianm actually the API endpoint is not changed. I only added a new config object for compaction config. Do you think the API endpoint should be changed as well?

gianm commented:
Sorry, I was being really dumb. I didn't realize the endpoint already existed through a different code path.

jihoonson commented:
@gianm oh, sorry. I finally got your point. Fixed docs according to the changed APIs.

jihoonson commented:
@gianm thanks. Fixed.
